users/tests.py

```diff
@@ -83,3 +84,18 @@ def test_filter_by_account_type(self, make_profile, user_idx, account_type, seco
         make_profile(second_user_idx, second_account_type)
         profiles_by_account_type = Profile.filter_by_account_type(new_profile.account_type)
         assert sorted(list(profiles_by_account_type)) == [new_profile]
+
+    @pytest.mark.parametrize("keyword_only_relevant_for_first_profile, account_type, num_of_profiles",
+                             [("first_name0", Account_type.STUDENT, 2),
+                              ("last_name0", Account_type.STUDENT, 2),
+                              ("last_name0", Account_type.STUDENT, 2)])
+    def test_search_users_by_keyword(self, make_profile, keyword_only_relevant_for_first_profile,
+                                     account_type, num_of_profiles):
+        profile_list = [make_profile(i, account_type) for i in range(num_of_profiles)]
```

Comment:
This test uses the `make_profile` fixture factory, so each profile generated will have unique fields in the format of `fieldname{i}` for `makeprofile(i,..)`
 
What I'm doing here is to create N profiles, then take the first profile with ind=0,
and for each field that is relevant for searching a user by a keyword,
it is asserted that only the first profile is in queryset, for a given `fieldname0` as a keyword. 

Since in this test i use the `make_profile` fixture factory i'm limited to "hardcoded" field names. Would you structure this test differently or add additional test in addition to this one?

Reply:
I don't understand what you mean by `it is asserted that only the first profile is in queryset`. can you give an example?

and a fixture (or a fixture factory) is a tool to help you, it should never limit you. if it does, you need to create a different fixture (it doesn't cost money :) )

Reply:
For example for query string `first_name0`, it is asserted that only the first profile (that was created with index 0) is in the queryset (whose `first_name` field is `first_name0`.
As opposed to profiles with indices `i=1,..,N` whose `first_name` field is `first_name1`, `first_name2`,...,`first_nameN`

Reply:
Maybe I'm missing something, so I'll try to clarify again: you check that the search function only finds the exact thing it searched for, meaning that when you search for `first_name0`, you won't find `first_name1` or `first_name` or `first_name01`?
does it cover all those cases? can we prove it?

Reply:
Yes exactly.

The way I implemented the search function is with `icontains` so case-insensitive containment for the query string test for all the relevant fields in user.
If for example given the 1,...n profiles  made with the make_profile factory i search for `first_name` or `First_NAME` i should get all the profiles in queryset.

Maybe I should also add two more tests to test case insensitive and string containment behavior?

Reply:
With tests for search or regex functions, you have to try and fail the test as much as possible, with multiple edge-cases.
So I'll test things like: Case-sensetive, typo, 1 character less than expected, 1 character more than expected. 

If time allows, try to achieve it today. it is important.

Reply:
Fixed

Reply:
Great work! well done